Initial support for FPGA CI Test Shield

fkjagodzinski opened the pull request:
- Bring `tools/fpga_ci_test_shield` from mbed-os @ 6.11.
- Enable `tests/mbed_hal_fpga_ci_test_shield/gpio`.
- Enable `tests/mbed_hal_fpga_ci_test_shield/gpio_irq`.
- A couple of updates to make this all work:
    - Update CMake files.
    - Disable all the _FPGA CI Test Shield_ features related to firmware handling (e.g. firmware update) with `EXTRA_DEPENDENCIES_REQUIRED__MBEDTESTER_FW_UPDATE` macro. These depend on mbed-os' `FileHandle` and `Callback`.
    - Bring `DigitalInOut` class from mbed-os @ 6.11 and make it part of `tools/fpga_ci_test_shield`.
    - Disable support for all peripherals except GPIO in `tools/fpga_ci_test_shield/include/fpga_ci_test_shield/test_utils.h`. This should be reverted/fixed when enabling more test-shield tests.
    - Disabled the C++ templates for atomic operations with a `EXTRA_DEPENDENCIES_REQUIRED__ATOMIC_TEMPLATES` macro. These templates require `mstd::type_identity` provided by the `mstd_type_traits` header. The standard `type_traits` is missing this feature for `cxx_std_14`.
    - Bring some of the sources back to build (e.g. `source/mbed_compat.c`, `source/mbed_pinmap_default.cpp`).
    - Use the custom IO greentea-client feature in test-shield tests.

fkjagodzinski commented:
This fixes #24 after the updates to `main` branch.

fkjagodzinski commented:
We already had a bit of discussion in the old PR. Here is the most relevant part https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/24#issuecomment-878204713:

> > The CMake name of the FPGA CI Test Shield lib, mbed-fpga-ci-test-shield -- do we keep that/do we rename now/do we rename as a part of bigger overhaul?
> 
> Keep it for now- we'll handle a rename later, as it impacts many drivers/files.
> 
> > The dir name tools/COMPONENT_FPGA_CI_TEST_SHIELD looks weird in this repo; we tend to use lower case -- shall I rename this to something like tools/fpga_ci_test_shield?
> 
> Yes please, but again, can be done later if it impacts many drivers/files.
> 
> > Is the tools/cxxsupport the right location for cxxsupport mbed-core feature?
> 
> IMO it's not the ideal place, but we need a restructure of files/folders anyway (see tools, bootstrap, hal/utils)
> I'd suggest we keep this here for now. @hugueskamba feel free to chime in
> 
> > Can we keep hal/source/mbed_pinmap_default.cpp in it's current location? Should I move that to tools/COMPONENT_FPGA_CI_TEST_SHIELD?
> 
> This seems needed regardless on whether there is a FPGA CI test shield or not.
> I'd suggest we rename it to `.c` as doesn't seem C++ code.
> 
> > Do we change the old-school SPI master/slave names already or shall this part wait for being actually used by the tests (when these get enabled)?
> 
> We should be able to change comments / docs as needed. However API needs to remain a bit more, so we don't break backwards compatibility. We'll address this as part of HAL API evolution.

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/CMakeLists.txt:
```diff
@@ -0,0 +1,20 @@
+# Copyright (c) 2020 ARM Limited. All rights reserved.
```
Please update header with new year

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/DigitalInOut.h:
```diff
@@ -0,0 +1,161 @@
+/* mbed Microcontroller Library
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/DynamicPinList.h:
```diff
@@ -0,0 +1,92 @@
+/*
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/I2CTester.h:
```diff
@@ -0,0 +1,187 @@
+/*
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/MbedTester.h:
```diff
@@ -0,0 +1,934 @@
+/*
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/SPIMasterTester.h:
```diff
@@ -0,0 +1,123 @@
+/*
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/SPISlaveTester.h:
```diff
@@ -0,0 +1,121 @@
+/*
```
Please update header with new year and license

MarceloSalazar commented on tools/COMPONENT_FPGA_CI_TEST_SHIELD/include/fpga_ci_test_shield/SPITester.h:
```diff
@@ -0,0 +1,127 @@
+/*
```
Please update header with new year and license

fkjagodzinski commented:
A couple of updates:

- Got rid of `tools/cxxsupport` (previously brought from mbed-os @ 6.11.) and disabled the C++ templates for atomic operations with a `USE_MBED_ATOMIC_TEMPLATES` macro. These templates require `mstd::type_identity` provided by the `mstd_type_traits` header. The standard `type_traits` is missing this feature for `cxx_std_14`. There is an alternative -- changing https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/blob/c6950a8d16a64f0f4cc874199609cda7eec19d0e/CMakeLists.txt#L99 to `cxx_std_20`. Any votes on going this way?
- Renamed `tools/COMPONENT_FPGA_CI_TEST_SHIELD` to `tools/fpga_ci_test_shield`.
- Updated all license headers in `tools/fpga_ci_test_shield`.

ithinuel commented on CMakeLists.txt:
```diff
@@ -166,23 +166,33 @@ add_subdirectory(examples/us_ticker_delay EXCLUDE_FROM_ALL)
-add_subdirectory(tests/mbed_hal/lp_ticker EXCLUDE_FROM_ALL)
-add_subdirectory(tests/mbed_hal/us_ticker_lp_ticker_common EXCLUDE_FROM_ALL)
-add_subdirectory(tests/mbed_hal/us_ticker_lp_ticker_frequency EXCLUDE_FROM_ALL)
+if (BUILD_GREENTEA_TESTS)
```
This changes the work flow and requires an extra argument to be passed to `cmake`.
Can we not have this for now or update the relevant documents?

MarceloSalazar replied:
Can we define this at the top level application/test's cmake configuration?
Some libraries may need to be compiled differently depending on this flag.

fkjagodzinski replied:
> or update the relevant documents?

Already done for ST in a pending PR https://github.com/MCU-Driver-HAL/MCU-Driver-ST/pull/15/commits/9bc4df16d3998e394c401005507776d492a8ac10.

I think the main cause for adding this arg was that the previous version of this change set added the `cxxsupport` that extended the `mbed-core` CMake target.

The above is no longer true but I still like the separation the arg adds to HAL and greentea tests. I'm fine with either keeping it or removing?

@MarceloSalazar keep / remove / update?

MarceloSalazar replied:
I'd prefer us to reduce the number of arguments and have consistency on how we compile apps / tests.
If we're compiling a test, the top level config should identify that that pass the flag accordingly, but don't know if that's possible. This should help us to compile all cmake projects from root dir

@ithinuel please review and share your thoughts

ithinuel replied:
the greentea tests and the hal are already nicely separated (in different cmake libraries) so this adds more complexity than effectively required.

ithinuel replied:
I vote for **not** having this flag!

fkjagodzinski replied:
All clear! I'll remove it.

MarceloSalazar replied:
I agree - we probably don't want to include code/libs that we don't need.
I assume we compile everything and the linker will remove what's not needed?
Do you have other solution in mind?

fkjagodzinski replied:
Done, 2e481b1

ithinuel commented on tools/fpga_ci_test_shield/include/fpga_ci_test_shield/DigitalInOut.h:
```diff
@@ -0,0 +1,160 @@
+/* Copyright (c) 2006-2021 Arm Limited
```
This class is pretty shallow and does not seem to be inherited anywhere.
Shall we take this opportunity **not** to add it and update the mbed tester to rely on gpio_api instead?

fkjagodzinski replied:
This still requires non-zero effort on updating the `MbedTester`. Will look into this.

ithinuel replied:
Yes, It was not a definitive request, but rather a question for @MarceloSalazar 's opinion.

MarceloSalazar replied:
I see there are more files / peripherals that would require an update to be consistent, hence I don't think we should do this right now. Let's enable tests and work on a clean up soon.

ithinuel commented on hal/source/mbed_gpio_irq.c:
```diff
@@ -14,12 +14,11 @@
 
-#include "platform/mbed_toolchain.h"
-#include "hal/gpio_api.h"
+#include "bootstrap/mbed_toolchain.h"
```
Is there a reason for having some include split in and out of the guarded region?
If not, can be keep them packed together (either in or out)?

fkjagodzinski replied:
Done https://github.com/MCU-Driver-HAL/MCU-Driver-HAL/pull/32/commits/c26369525f2218d55e4d16568a9468363fb5c0a4

ithinuel commented on tests/mbed_hal_fpga_ci_test_shield/gpio/main.cpp:
```diff
@@ -20,14 +20,16 @@
 
-#include "utest/utest.h"
-#include "unity/unity.h"
+#include "./gpio_fpga_test.h"
```
why do we have a relative include path here?

fkjagodzinski replied:
Oh, this is just an effect of me really getting to like the ordered headers list. The effect I really like is that related headers are grouped (by dir prefix). I wanted to have that feature also for the closest ones. Of course this might make more sense when we have more headers in `./`.
Anyway, do we want to introduce such convention or shall I change that to `#include "gpio_fpga_test.h"`

MarceloSalazar replied:
I'd suggest we keep consistency.
I haven't seen relative paths elsewhere, so I'd propose we use `#include "gpio_fpga_test.h"` if that's ok?

ithinuel replied:
This is just an opinion but I tend to prefer absolute path to serve as fully qualified namespaces for headers.
This proved to me more reliable and less error prone than relative path with possibly colliding file names.
This also simplifies the definition of include path with just the root of the module/project.

fkjagodzinski replied:
Fixed via ac1cbae & 544ac39.

ithinuel commented on tests/mbed_hal_fpga_ci_test_shield/gpio_irq/main.cpp:
```diff
@@ -22,19 +22,19 @@
+#include "fpga_ci_test_shield/MbedTester.h"
+#include "fpga_ci_test_shield/test_utils.h"
 #include "gpio_irq_fpga_test.h"
+#include "hal/gpio_api.h"
```
This breaks the alphabetical ordering of the includes.

ithinuel commented on tools/fpga_ci_test_shield/CMakeLists.txt:
```diff
@@ -0,0 +1,32 @@
+target_include_directories(mbed-fpga-ci-test-shield
+    INTERFACE
+        include
+        include/fpga_ci_test_shield
```
We should not add that line to enforce the benefit of using include path as namespaces.

fkjagodzinski replied:
Done c92bd5d8bf4f125bec977f7dc63317e84d00248a

fkjagodzinski commented:
Latest updates:

- Changed `hal/source/mbed_pinmap_default.cpp` to `.c` and added changes required to fix `error: initializer element is not constant` (C vs C++ difference).
- Fixed a couple of issues with `#include` directives.
- Renamed the macros I previously added to disable some features -- added a common prefix:
    - `EXTRA_DEPENDENCIES_REQUIRED__ATOMIC_TEMPLATES`,
    - `EXTRA_DEPENDENCIES_REQUIRED__MBEDTESTER_FW_UPDATE`.

I'll now work on removing `DigitalInOut` class. Apart from that, this is read to be re-reviewed.

ithinuel commented on hal/CMakeLists.txt:
```diff
@@ -23,14 +23,14 @@ target_include_directories(mbed-core
         # source/mbed_itm_api.c
         source/mbed_pinmap_common.c
-        # source/mbed_pinmap_default.cpp
+        source/mbed_pinmap_default.cpp
```
Thinking about it, those rely heavily on `weak` symbols and I would love so see them annihilated.
We could get rid of `MBED_WEAK` in there and gate those function with an `#ifndef CONFIG_NO_DEFAULT_PINMAP` when a target wants to define its own pinmap.

OTOH this kind of softly break the promise of backward compat with Mbed-OS 6.11 compat'.
It is still compatible but requires an extraflag but we already break that promise by using cmake rather than `mbed-cli` so 🤷 .

@MarceloSalazar thoughts?

MarceloSalazar replied:
This is something that has been identified as dependency  and the need to clean/fix, but not at the moment (requires an update of all MCU-Driver-HAL)

MarceloSalazar commented:
@fkjagodzinski let us know when the PR is ready for re-review

fkjagodzinski commented:
> @fkjagodzinski let us know when the PR is ready for re-review

It is now! :smiley: 

I'll clean up/squash the commit history once this is accepted. This will also remove the commit that didn't pass the DCO check.

ithinuel commented on hal/source/mbed_pinmap_default.c:
```diff
@@ -91,18 +91,18 @@ MBED_WEAK const PinList *pinmap_gpio_restricted_pins()
+
+static PeripheralList restricted_uart_peripheral_list = {
+    .count = 0,
+    .peripheral = restricted_uart_peripherals
```
Does this pointer hast to be valid if `count = 0`?
It seems fair to me to have a NULL pointer there and simplify the code :)

ithinuel replied:
Does this pointer hast to be valid if `count = 0`?
It seems fair to me to have a NULL pointer there and simplify the code :)

ithinuel reviewed (commented):
~Almost there :D Last one is more a question/comment than a request for changes.~

A nice squash/rebase into a few sensible commits and we'll be ready to merge 🎉

fkjagodzinski commented:
Rebase done. I managed to reduce the number of commits down to 8! @MarceloSalazar , @ithinuel your reviews got cancelled by that force push, so I have to ask you to review/accept again. :wink:
